### ISSUE TITLE
💅 `useExhaustiveDependencies` should not warn on a function wrapped in `useCallback` has a recursive call within

### ISSUE THREAD

**Issue by misicnenad** [biomejs/biome#2361]
### Environment information

_Note: the same bug exists in biome@1.6.4._
```bash
CLI:
  Version:                      1.5.3
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           linux

Environment:
  BIOME_LOG_DIR:                unset
  NO_COLOR:                     unset
  TERM:                         "xterm-256color"
  JS_RUNTIME_VERSION:           "v19.6.0"
  JS_RUNTIME_NAME:              "node"
  NODE_PACKAGE_MANAGER:         "npm/9.4.0"

Biome Configuration:
  Status:                       Loaded successfully
  Formatter disabled:           false
  Linter disabled:              false
  Organize imports disabled:    false
  VCS disabled:                 true

Workspace:
  Open Documents:               0
```

### Rule name

useExhaustiveDependencies

### Playground link

https://biomejs.dev/playground/?lintRules=all&code=aQBtAHAAbwByAHQAIAB7ACAAdQBzAGUAQwBhAGwAbABiAGEAYwBrACAAfQAgAGYAcgBvAG0AIAAiAHIAZQBhAGMAdAAiADsACgAKAGUAeABwAG8AcgB0ACAAZgB1AG4AYwB0AGkAbwBuACAAQwBvAG0AcABvAG4AZQBuAHQAKAApACAAewAgACAACgAgACAAYwBvAG4AcwB0ACAAZgBpAGIAIAA9ACAAdQBzAGUAQwBhAGwAbABiAGEAYwBrACgAKABuAHUAbQA6ACAAbgB1AG0AYgBlAHIAKQA6ACAAbgB1AG0AYgBlAHIAIAA9AD4AIAB7AAoAIAAgACAAIABpAGYAIAAoAG4AdQBtACAAPAAgADIAKQAgAHsACgAgACAAIAAgACAAIAByAGUAdAB1AHIAbgAgAG4AdQBtADsACgAgACAAIAAgAH0ACgAgACAAIAAgAHIAZQB0AHUAcgBuACAAZgBpAGIAKABuAHUAbQAgAC0AIAAxACkAIAArACAAZgBpAGIAKABuAHUAbQAgAC0AIAAyACkAOwAKACAAIAB9ACwAIABbAF0AKQA7AAoACgAgACAAcgBlAHQAdQByAG4AIAA8AD4APAAvAD4ACgB9AA%3D%3D

### Expected result

It should not throw any error when a function wrapped in `useCallback` calls itself.

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct
